wdl-lint/src/rules/redundant_input_assignment.rs

```diff
@@ -0,0 +1,90 @@
+//! A lint rule for flagging redundant input assignments
+
+use std::fmt::Debug;
+use wdl_ast::{AstToken, VersionStatement};
+use wdl_ast::Diagnostic;
+use wdl_ast::Diagnostics;
+use wdl_ast::Document;
+use wdl_ast::Span;
+use wdl_ast::SupportedVersion;
+use wdl_ast::v1::{Decl, InputSection};
+use wdl_ast::VisitReason;
+use wdl_ast::Visitor;
+
+use crate::Rule;
+use crate::Tag;
+use crate::TagSet;
+
+/// The identifier for the Redundant Input Assignment rule.
+const ID: &str = "RedundantInputAssignment";
+
+/// Create a "Redundant Input Assignment" diagnostic.
+fn redundant_input_assignment(span: Span) -> Diagnostic {
+    Diagnostic::note("redundant input assignments can be shortened")
+        .with_rule(ID)
+        .with_label("redundant input assignments can be shortened", span)
+}
+
+/// Detects a malformed lint directive.
+#[derive(Default, Debug, Clone, Copy)]
+pub struct RedundantInputAssignment {
+    included_version: bool,
+}
+
+impl Rule for RedundantInputAssignment {
+    fn id(&self) -> &'static str {
+        ID
+    }
+
+    fn description(&self) -> &'static str {
+        "Flags redundant input assignments."
+    }
+
+    fn explanation(&self) -> &'static str {
+        "Input assignments that are redundant can be shortened. For example, `{ input: a = a }` \
+        can be shortened to `{ input: a }`."
+    }
+
+    fn tags(&self) -> TagSet {
+        TagSet::new(&[Tag::Clarity, Tag::Correctness])
+    }
+
+    fn exceptable_nodes(&self) -> Option<&'static [wdl_ast::SyntaxKind]> {
+        None
+    }
+}
+
+impl Visitor for RedundantInputAssignment {
+    type State = Diagnostics;
+
+    fn document(&mut self, _: &mut Self::State, _: VisitReason, _: &Document, _: SupportedVersion) {
+        // This is intentionally empty, as this rule has no state.
+    }
+
+    fn version_statement(&mut self, state: &mut Self::State, reason: VisitReason, stmt: &VersionStatement) {
+        // this current takes into account for any possible patch versions, which might be unnecessary
+        let binding = stmt.version();
+        let split_versions = binding.as_str().split(".").collect::<Vec<_>>();
+        if split_versions[0] == "1" && split_versions[1] == "0" {
+            self.included_version = false;
+        } else {
+            self.included_version = true;
+        }
+    }
+
+    fn input_section(&mut self, state: &mut Self::State, reason: VisitReason, section: &InputSection) {
+        /// if the version statement is earlier than 1.1 return
+        if !self.included_version {
+            return
+        }
+
+        /// for each declaration in the input section,
+        section.declarations().for_each(|dcl| {
+            if let Some(expr) = dcl.expr() {
+                // if dcl.name().as_str() == expr {
+                //     state.add(redundant_input_assignment(Span::new(0, 12)));
+                // }
```

Comment:
@a-frantz this is where I'm hitting my confusion. I'm not sure the best way to turn the `expr` into a string for direct comparison, and have been hitting some difficulties understanding how to access the span for each line of `input_section`

Reply:
I think you can get this done without ever calling `as_str()`. A few scattered thoughts here:
- `as_str()` is actually quite expensive to call _for nodes_ (and `Expr` is a node)
    - you should try to only call `as_str()` once you've parsed down to a token, which is cheap
    - It's expensive for nodes because it involves traversing the CST to visit every child node and calling `as_str()` on them
- `if let Some(expr) = dcl.expr()` this is a good start. In the case that `.expr()` returns none, that means the declaration is already using the abbreviated syntax you're checking for. So this part looks good 👍 
- Next you want to check what kind of expression is present. You only care about when it is a `NameRef`
    - https://docs.rs/wdl/latest/wdl/ast/v1/enum.Expr.html#variant.Name
    - So check if it can be ceorced to a `NameRef`, and if so check if the name is equivalent to the `dcl.name()`
    - I could be wrong, but I think that check can be done be comparing the tokens to each other (without converting to str)

Reply:
O and for the span: I think you want to highlight the entire declaration, so you can do something like `dcl.to_span()` instead of trying to construct a new span yourself.

Reply:
Er. Actually. In light of #243 I would just highlight the `Expr`/`NameRef`